### ISSUE TITLE
Including f0 in ShimOpt.optimizeshimcurrents()

### ISSUE THREAD

**Issue by rtopfer** [shimming-toolbox/shimming-toolbox-matlab#33]
the transmit frequency f0 should be included in the shim optimization (it's the simplest yet most important term), yet the current code yields buggy results when it is allowed to vary from its original setting (i.e. the imaging frequency of the input field map)...

**Comment by jaystock** [comment]
For multicoil shim arrays, does the shim optimizer allow the maximum current limit to be specified separately for each coil channel?   For some shim arrays, I could envision wanting to impose a different current limit for each coil.

**Comment by rtopfer** [comment]
> For multicoil shim arrays, does the shim optimizer allow the maximum current limit to be specified separately for each coil channel? 

Indeed, each shim system has default limits set in a parameter file:
Currently this is a classdef for each system (e.g. [8ch neck coil](https://github.com/shimming-toolbox/shimming-toolbox/blob/0f47f42054f84caff85f6d59c6499ad182590e01/Coils/Shim_Greg/ShimSpecs_Greg.m#L63))
But I'm suggesting in [issue75](https://github.com/shimming-toolbox/shimming-toolbox/issues/75) that these ShimSpecs files just be .json text files instead.

@jaystock I was wondering whereabouts in your code you specify system parameters? Also, suspect that the existing specs aren't exhaustive and maybe you have others that should be included in a .json file?

Also, the current limits given in the Specs files are treated as defaults for the optimization: 
The user can reconfigure channel limits with an input `Params` (struct field) in `ShimOpt.optimizeshimcurrents`: 
https://github.com/shimming-toolbox/shimming-toolbox/blob/0f47f42054f84caff85f6d59c6499ad182590e01/Coils/Shim_Template/ShimOpt.m#L1040

**Comment by jaystock** [comment]
Thank you @rtopfer for clarifying.   Not to split hairs, but is there also a way to set which shim channels count toward the overall total current limit?   My reason for asking is that when I jointly optimize the Siemens shims and the multi-coil shims together (to get a slight performance gain vs. optimizing sequentially), I have a total current limit for the multi-coil, and a channel-wise limit for each Siemens shim term.  But the Siemens shim currents don't count toward the "Total Current" limit, which only applies to the multi-coil shims.

Which system parameters are you thinking about?  We usually include the current limits for our multicoil shims in the Matlab data structure that we read in from a .mat file.  Likewise for the Siemens shims.  I like to store the metadata in a .mat file for the Siemens shim terms; I also like to use a point-by-point B0 field map to compute the Siemens shims instead of using pure 1st and 2nd order fields.  This accounts for imperfections and cross-terms.  I have been getting excellent agreement between my offline Siemens shim calculations and acquired B0 maps (better than I do for my multi-coil shims actually).  But a json file with the hardware specifications probably makes more sense than just stuffing that info into a .mat file.

**Comment by rtopfer** [comment]
> is there also a way to set which shim channels count toward the overall total current limit?

@jaystock Sorry for the long-winded answer but the way this is dealt with in the current code is sort of esoteric: 

Basically, the is a generic `ShimOpt.optimizeshimcurrents()` method which, as it stands, doesn't limit "total current"—it's strictly per-channel. 

However, each shim subsystem can _optionally_ redefine their own version, e.g.: For the 24-channel RRI coil, there _is_ a total current limit as well as per-bank limit on current (24-channels are allocated to amplifier 4 banks): This gets defined as a _local function_ in its custom method `ShimOpt_Rriyan.optimizeshimcurrents()` 
https://github.com/shimming-toolbox/shimming-toolbox/blob/0f47f42054f84caff85f6d59c6499ad182590e01/Coils/Shim_Rriyan/ShimOpt_Rriyan.m#L86 

To avoid fully duplicating the rest of the optimization code (apart from this minor distinction around current constraints, the rest of the optimization is the same irrespective of the specific coil) this custom constraint function simply gets passed as an argument (function handle) to the generic method, i.e. the `@checknonlinearconstraint` 3rd argument here:
https://github.com/shimming-toolbox/shimming-toolbox/blob/0f47f42054f84caff85f6d59c6499ad182590e01/Coils/Shim_Rriyan/ShimOpt_Rriyan.m#L83

(Maybe worth noting that a `ShimOpt` subclass for a specific system isn't obligated to call on the generic optimization method at all, it can redefine everything entirely; in the end, it just needs to spit out a vector of correction currents.)



 
_2nd point:_
>
> Which system parameters are you thinking about? We usually include the current limits for our multicoil shims in the Matlab data structure that we read in from a .mat file. Likewise for the Siemens shims. 

Exactly that sort of thing. I include everything needed to describe a given system in a single [ShimSpecs](https://github.com/shimming-toolbox/shimming-toolbox/blob/953f4bbc69886049c7054a23ea9ce8511db349d0/Coils/Shim_Greg/ShimSpecs_Greg.m) file (essentially it's just a Matlab struct rather than a class, which is why I feel it may as well just be stored a .json file). Some of the "parameters" are merely cosmetic things like "Id" (e.g. "Greg"); some parameters (like baudrate) are used exclusively for communicating with the hardware through Matlab; others, like `maxCurrentPerChannel` are used both in the optimization and hardware communication.

My feeling is that other systems might have additional parameters that would need to be specified in a standard .json, but maybe not.

> I also like to use a point-by-point B0 field map to compute the Siemens shims instead of using pure 1st and 2nd order fields. This accounts for imperfections and cross-terms. I have been getting excellent agreement between my offline Siemens shim calculations and acquired B0 maps (better than I do for my multi-coil shims actually).

Indeed, I redefine separate classes for the "ideal" (simulated) shims and the empirically mapped ones, but maybe there's a nicer/more concise way of bundling them together...
